homeassistant/components/upb/light.py

```diff
@@ -0,0 +1,105 @@
+"""Platform for UPB light integration."""
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_FLASH,
+    ATTR_TRANSITION,
+    SUPPORT_BRIGHTNESS,
+    SUPPORT_FLASH,
+    SUPPORT_TRANSITION,
+    Light,
+)
+from homeassistant.helpers import entity_platform
+
+from . import UpbAttachedEntity
+from .const import DOMAIN, UPB_BLINK_RATE_SCHEMA, UPB_BRIGHTNESS_RATE_SCHEMA
+
+SERVICE_LIGHT_FADE_START = "light_fade_start"
+SERVICE_LIGHT_FADE_STOP = "light_fade_stop"
+SERVICE_LIGHT_UPDATE_STATUS = "light_update_status"
+SERVICE_LIGHT_BLINK = "light_blink"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the UPB light based on a config entry."""
+
+    upb = hass.data[DOMAIN]["upb"]
+    async_add_entities(UpbLight(upb.devices[dev], upb) for dev in upb.devices)
+
+    platform = entity_platform.current_platform.get()
+
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_FADE_START, UPB_BRIGHTNESS_RATE_SCHEMA, "light_fade_start"
+    )
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_FADE_STOP, {}, "light_fade_stop"
+    )
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_UPDATE_STATUS, {}, "light_update_status"
+    )
+    platform.async_register_entity_service(
+        SERVICE_LIGHT_BLINK, UPB_BLINK_RATE_SCHEMA, "light_blink"
+    )
+
+
+class UpbLight(UpbAttachedEntity, Light):
+    """Representation of an UPB Light."""
+
+    def __init__(self, element, upb):
+        """Initialize an UpbLight."""
+        super().__init__(element, upb)
+        self._brightness = self._element.status
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        if self._element.dimmable:
+            return SUPPORT_BRIGHTNESS | SUPPORT_TRANSITION | SUPPORT_FLASH
+        return SUPPORT_FLASH
+
+    @property
+    def brightness(self):
+        """Get the brightness."""
+        return self._brightness
+
+    @property
+    def is_on(self) -> bool:
+        """Get the current brightness."""
+        return self._brightness != 0
+
+    async def async_turn_on(self, **kwargs):
+        """Turn on the light."""
+        flash = kwargs.get(ATTR_FLASH)
+        if flash:
+            self.light_blink(0.5 if flash == "short" else 1.5)
+        else:
+            rate = kwargs.get(ATTR_TRANSITION, -1)
+            brightness = kwargs.get(ATTR_BRIGHTNESS, 255) / 2.55
+            self._element.turn_on(brightness, rate)
```

Comment:
No I/O here right?

Reply:
Not right. `turn_on` ends up doing async io. Why?

Reply:
Specifically I noticed the lack of `await` so I am checking for blocking I/O in async context to make sure the event loop isn't blocked.   For more context:

https://www.home-assistant.io/blog/2020/04/29/release-109/#detection-of-blocking-io-in-the-event-loop

Reply:
Understand. Thx. No blocking IO -- entire lib is asyncio.